docs/site/tutorials/protocol_oriented_generics.ipynb

```diff
@@ -0,0 +1,1307 @@
+        "\n",
+        "let products = [\n",
+        "    Laptop(name: \"a\", box: .small, mass: .light),\n",
+        "    Laptop(name: \"b\", box: .large, mass: .medium),\n",
+        "    Laptop(name: \"c\", box: .medium, mass: .heavy),\n",
+        "    Laptop(name: \"d\", box: .large, mass: .light)\n",
+        "]\n",
+        "\n",
+        "print(filtering(servers, by: BoxFilter(box: .small)))\n",
+        "print(filtering(servers, by: MassFilter(mass: .medium)))\n",
+        "\n",
+        "print(filtering(products, by: BoxFilter(box: .small)))\n",
+        "print(filtering(products, by: MassFilter(mass: .medium)))"
+      ],
+      "execution_count": 17,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "text": [
+            "[(working serverA small heavy), (working serverE small heavy)]\r\n",
+            "[(notWorking serverB medium medium)]\r\n",
+            "[(a small light)]\r\n",
+            "[(b large medium)]\r\n"
+          ],
+          "name": "stdout"
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "colab_type": "text",
+        "id": "-UrRDSaFNCRg"
+      },
+      "source": [
+        "We have now been able to filter an array by not only any property of a specific `struct`, but also be able to filter any struct which has that property!"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "6g3pPZIuMvPu",
+        "colab_type": "text"
+      },
+      "source": [
+        "# Tips for good API design\n",
+        "***This section was taken from the [WWDC 2019: Modern Swift API Design](https://developer.apple.com/videos/play/wwdc2019/415/) talk.***\n",
+        "\n",
+        "Now that you understand how protocols behave, it's best to now go over when you should use protocols. As powerful as protocols can be, it's not always the best idea to dive in and immediately start with protocols.\n",
+        "\n",
+        "* Start with concrete use cases.\n",
+        "    * First explore the use case with concrete types and understand what code it is you want to share and find is being repeated. Then, factor that shared code out with generics. It might mean to create new protocols. Discover a need for generic code.\n",
+        "* Consider composing new protocols from existing protocols defined in the standard library. Refer to the following [Apple documentation](https://developer.apple.com/documentation/swift/adopting_common_protocols) for a good example of this.\n",
+        "* Instead of a generic protocol, consider defining a generic type instead.\n",
+        "\n",
+        "## Example: defining a custom vector type\n",
+        "Let's say we want to define a `GeometricVector` protocol on floating-point numbers to use in some geometry app we are making which defines 3 important vector operations:\n",
+        "\n",
+        "```swift\n",
+        "protocol GeometricVector where Scalar: FloatingPoint {\n",
```

Comment:
Move the constraint in the where clause to the declaration of the associated type scalar.